src/keystore/InviteStore.ts

```diff
@@ -0,0 +1,76 @@
+import { keystore } from '@xmtp/proto'
+import { Persistence } from './persistence/interface'
+import type { TopicData } from './interfaces'
+import { Mutex } from 'async-mutex'
+import { typeSafeTopicMap } from './utils'
+
+const INVITE_KEY = 'invitations:v1'
+
+/**
+ * InviteStore holds a simple map of topic -> TopicData and writes to the persistence layer on changes
+ */
+export default class InviteStore {
+  private persistence?: Persistence
+  private mutex: Mutex
+  private topicMap: Map<string, TopicData>
+
+  constructor(
+    persistence?: Persistence,
+    initialData: Map<string, TopicData> = new Map()
+  ) {
+    this.persistence = persistence
+    this.mutex = new Mutex()
+    this.topicMap = initialData
+  }
+
+  static async create(persistence?: Persistence): Promise<InviteStore> {
+    if (persistence) {
+      const rawData = await persistence.getItem(INVITE_KEY)
+      if (rawData) {
+        try {
+          const inviteMap = typeSafeTopicMap(keystore.TopicMap.decode(rawData))
+          // Create an InviteStore with data preloaded
+          return new InviteStore(
+            persistence,
+            new Map(Object.entries(inviteMap))
+          )
+        } catch (e) {
+          console.warn(`Error loading invites from store: ${e}`)
+        }
+      }
+    }
+    return new InviteStore(persistence)
+  }
+
+  async add(topicData: TopicData[]): Promise<void> {
+    await this.mutex.runExclusive(async () => {
+      let isDirty = false
+      for (const row of topicData) {
+        // This will not overwrite any existing values. First invite found in the store for a given topic will always be used
+        // Duplicates do not throw errors
+        if (!this.topicMap.has(row.invitation.topic)) {
```

Comment:
Is this a necessity? or simplification of edge cases for now?

Reply:
In cases where there are multiple invites for the same topic, we do need to decide which one to use. Oldest wins seems like the correct call.

It removes an attack vector where someone can know another user's topic and "poison" it by sending them another invite using the same topic but a different encryption key.

So long as invites are loaded in ascending order, this should do that.